src/DataStructures/FixedHashMap.hpp

```diff
@@ -0,0 +1,568 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <array>
+#include <boost/none.hpp>
+#include <boost/optional.hpp>
+#include <cstddef>
+#include <iterator>
+#include <pup.h>
+#include <pup_stl.h>
+#include <type_traits>
+
+#include "ErrorHandling/Assert.hpp"
+#include "Parallel/PupStlCpp11.hpp"  // IWYU pragma: keep
+#include "Utilities/Algorithm.hpp"
+#include "Utilities/BoostHelpers.hpp"  // IWYU pragma: keep
+#include "Utilities/ConstantExpressions.hpp"
+#include "Utilities/ForceInline.hpp"
+#include "Utilities/GetOutput.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/Numeric.hpp"
+#include "Utilities/PrintHelpers.hpp"
+#include "Utilities/TypeTraits.hpp"
+
+/// \cond
+template <size_t MaxSize, class Key, class ValueType, class Hash,
+          class KeyEqual>
+class FixedHashMapIterator;
+/// \endcond
+
+namespace FixedHashMap_detail {
+template <typename T, typename = cpp17::void_t<>>
+struct has_is_perfect_member : std::false_type {};
+
+template <typename T>
+struct has_is_perfect_member<T,
+                             cpp17::void_t<decltype(T::template is_perfect<0>)>>
+    : std::true_type {};
+
+template <typename T>
+constexpr bool has_is_perfect_member_v = has_is_perfect_member<T>::value;
+
+template <class T, size_t MaxSize,
+          bool HasIsPerfect = has_is_perfect_member_v<T>>
+constexpr bool is_perfect = T::template is_perfect<MaxSize>;
+
+template <class T, size_t MaxSize>
+constexpr bool is_perfect<T, MaxSize, false> = false;
+}  // namespace FixedHashMap_detail
+
+/*!
+ * \ingroup DataStructuresGroup
+ * \brief A hash table with a compile-time specified maximum size and ability to
+ * efficiently handle perfect hashes.
+ *
+ * There are a few requirements on the types passed to `FixedHashMap`. These
+ * are:
+ * - `Key` is CopyConstructible
```

Comment:
why CopyConstructible?

Reply:
Because we have a `std::pair<const key_type, mapped_type>` the `const` makes copy, move, and serialization all require the key type is copy constructible. I think with `const_cast` we can make move semantics work for non-copyable keys, but I'm not sure how easy serialization would be (I think this might require Charm++ modifications). @wthrowe @kidder would you like me to enable move semantics for non-copyable keys?

Reply:
I would say only if we need it

Reply:
:+1: I just tested `std::unordered_map` and it also requires keys be copy constructible for copy and move semantics (I didn't test serialization, but that's a Charm++ thing anyway)